### PR TITLE
Don't read lowering options from hidden global options

### PR DESCRIPTION
**Pull request by darthscsi**
Add a pass to overwrite lowering options on a module.  This lets us get rid of the in-library cmdline option and move it to the two tools which need it.  With this, prepare and export never change the lowering options.

Also make prepare a nested pass.

### CODE REVIEW

File: include/circt/Conversion/ExportVerilog.h [hunk 1]
```diff
@@ -13,12 +13,13 @@
 #ifndef CIRCT_TRANSLATION_EXPORTVERILOG_H
 #define CIRCT_TRANSLATION_EXPORTVERILOG_H
 
+#include "circt/Support/LoweringOptions.h"
```
Comment (dtzSiFive) [inline include/circt/Conversion/ExportVerilog.h:16]:
```suggestion
```

Comment (darthscsi) [reply include/circt/Conversion/ExportVerilog.h:16]: Shoot, I had deleted that in the branch before the merge.

File: lib/Conversion/ExportVerilog/PrepareForEmission.cpp [hunk 2]
```diff
@@ -1019,9 +1019,8 @@ void ExportVerilog::prepareHWModule(hw::HWModuleOp module,
...
 namespace {
 
-struct TestPrepareForEmissionPass
-    : public TestPrepareForEmissionBase<TestPrepareForEmissionPass> {
-  TestPrepareForEmissionPass() {}
+struct PrepareForEmissionPass
+    : public PrepareForEmissionBase<PrepareForEmissionPass> {
   void runOnOperation() override {
     HWModuleOp module = getOperation();
     LoweringOptions options = getLoweringCLIOption(
```
Comment (dtzSiFive) [inline lib/Conversion/ExportVerilog/PrepareForEmission.cpp:1026]: This means the options are taken from the CLI options only, not the module, is that correct?

Would be good to audit the plumbing of these options so we don't have any surprises, oof.

I think one result of this change (I think?) is any LoweringOptions on the module going into ExportVerilog will be ignored for this portion of the lowering (only using the CLI options), but not for the bits that grab them from the module (`LoweringOptions options(module)`)?  We do a `applyLoweringCLOptions` which IIRC merges them, but here this runs prepare only with the CLI options.  Generally I'd expect these to be the same, but they may not be-- for example, when saving a reproducer for example it's nice to not have to specify them if they're already on the module.

Mostly I'm concerned about ways that different flows might end up with /different/ options at various points.

File: lib/Conversion/ExportVerilog/ExportVerilog.cpp [hunk 3]
```diff
@@ -5001,26 +5002,14 @@ void SharedEmitterState::emitOps(EmissionList &thingsToEmit, raw_ostream &os,
   }
 }
 
-/// Prepare the given MLIR module for emission.
```
Comment (dtzSiFive) [inline lib/Conversion/ExportVerilog/ExportVerilog.cpp:5004]: Is there a reason to drop this? Looks like it'd be useful for two copies of its body below (modulo the `LoweringOptions options(module)` first).

File: lib/Conversion/ExportVerilog/ExportVerilog.cpp [hunk 4]
```diff
@@ -5001,26 +5002,14 @@ void SharedEmitterState::emitOps(EmissionList &thingsToEmit, raw_ostream &os,
...
 //===----------------------------------------------------------------------===//
 // Unified Emitter
 //===----------------------------------------------------------------------===//
 
-LogicalResult circt::exportVerilog(ModuleOp module, llvm::raw_ostream &os) {
-  // Prepare the ops in the module for emission and legalize the names that will
-  // end up in the output.
-  LoweringOptions options(module);
-  prepareForEmission(module, options);
+static LogicalResult exportVerilogImpl(ModuleOp module, llvm::raw_ostream &os) {
```
Comment (dtzSiFive) [inline lib/Conversion/ExportVerilog/ExportVerilog.cpp:5009]: So these "Impl's":
1) Assume/Require PrepareForEmission has run
2) Assume/Expect LoweringOptions that should be used have been attached to the module already.

Sounds good to me, just checking their usage/expectations.

File: lib/Conversion/ExportVerilog/ExportVerilog.cpp [hunk 5]
```diff
@@ -5066,7 +5064,14 @@ struct ExportVerilogPass : public ExportVerilogBase<ExportVerilogPass> {
     // TODO: This should be moved up to circt-opt and circt-translate.
     applyLoweringCLOptions(getOperation());
 
-    if (failed(exportVerilog(getOperation(), os)))
+    // Prepare the ops in the module for emission.
+    mlir::OpPassManager preparePM("builtin.module");
+    auto &modulePM = preparePM.nest<hw::HWModuleOp>();
+    modulePM.addPass(createPrepareForEmissionPass());
+    if (failed(runPipeline(preparePM, getOperation())))
+      signalPassFailure();
```
Comment (dtzSiFive) [inline lib/Conversion/ExportVerilog/ExportVerilog.cpp:5072]:
```suggestion
      return signalPassFailure();
```
(don't continue execution if there's an error)

File: lib/Conversion/ExportVerilog/ExportVerilog.cpp [hunk 6]
```diff
@@ -5210,6 +5224,14 @@ struct ExportSplitVerilogPass
     // on the command line.
     // TODO: This should be moved up to circt-opt and circt-translate.
     applyLoweringCLOptions(getOperation());
+
+    // Prepare the ops in the module for emission.
+    mlir::OpPassManager preparePM("builtin.module");
+    auto &modulePM = preparePM.nest<hw::HWModuleOp>();
+    modulePM.addPass(createPrepareForEmissionPass());
+    if (failed(runPipeline(preparePM, getOperation())))
+      signalPassFailure();
```
Comment (dtzSiFive) [inline lib/Conversion/ExportVerilog/ExportVerilog.cpp:5233]:
```suggestion
      return signalPassFailure();
```

File: lib/Conversion/ExportVerilog/ExportVerilog.cpp [hunk 7]
```diff
@@ -5210,6 +5224,14 @@ struct ExportSplitVerilogPass
...
+
     if (failed(exportSplitVerilog(getOperation(), directoryName)))
```
Comment (dtzSiFive) [inline lib/Conversion/ExportVerilog/ExportVerilog.cpp:5235]:
```suggestion
    if (failed(exportSplitVerilogImpl(getOperation(), directoryName)))
```
So the prepare isn't run twice.

**Review by dtzSiFive — COMMENTED** [review]
Great, look forward to this landing again soon!

Added some nits--primarily the handling of LoweringOptions seems a bit problematic, but TBH it's a bit messy already and I'm not sure how to best handle them.  Mostly want to make it hard/impossible to have "prepare" and "export" use different sets of options as that will likely cause problems.

In one place I believe this runs the preparation twice, see comment.

File: lib/Conversion/ExportVerilog/PrepareForEmission.cpp (hunk 2 above)
Comment (uenoku) [reply lib/Conversion/ExportVerilog/PrepareForEmission.cpp:1026]: > This means the options are taken from the CLI options only, not the module, is that correct?

This `getLoweringCLIOption` tries getting a lowering option from CLI option first, and then checks module's attribute. (Sorry for naming a confusing method, it should be renamed for more clear name)

File: lib/Conversion/ExportVerilog/ExportVerilog.cpp (hunk 4 above)
Comment (darthscsi) [reply lib/Conversion/ExportVerilog/ExportVerilog.cpp:5009]: They will use the default if there are no lowering options.  But yes.  Also will be broken out a bit better soon.

File: lib/Conversion/ExportVerilog/PrepareForEmission.cpp (hunk 2 above)
Comment (darthscsi) [reply lib/Conversion/ExportVerilog/PrepareForEmission.cpp]: You make me rethink all the lowering options.

Comment (youngar) [reply lib/Conversion/ExportVerilog/PrepareForEmission.cpp]: Just making sure I got this right:  I think the original plan was that the command line options would be set on the module (by `firtool`), and then all passes would read the options from there. I guess that missed the use-case of `circt-opt` wanting to take advantage of the CLI options, and suddenly we have passes reading CLI options directly, which we don't think is a great idea.  We are moving `firtool` back to the original design, but adding a new utility pass to `circt-opt` to apply command line options to the module.  Due to the main functionality of `circt-opt` living in MLIR core, we don't have the option to add the same command line options there.

File: include/circt/Support/LoweringOptionsParser.h [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+  bool parse(llvm::cl::Option &option, StringRef argName, StringRef argValue,
+             LoweringOptions &value) {
+    bool failed = false;
+    value.parse(argValue, [&](Twine error) { failed = option.error(error); });
+    return failed;
+  }
+};
+
+struct loweringOptionsOption
```
Comment (youngar) [inline include/circt/Support/LoweringOptionsParser.h:37]: Should this be lower case?

File: include/circt/Support/LoweringOptionsParser.h [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+                "disallowLocalVariables, verifLabels, emittedLineLength=<n>, "
+                "maximumNumberOfTermsPerExpression=<n>, "
+                "maximumNumberOfTermsInConcat=<n>, explicitBitcast, "
+                "maximumNumberOfVariadicOperands=<n>, "
+                "emitReplicatedOpsToHeader, "
+                "locationInfoStyle={plain,wrapInAtSquareBracket,none}, "
+                "disallowPortDeclSharing, printDebugInfo, useOldEmissionMode, "
+                "disallowExpressionInliningInPorts"),
+            llvm::cl::cat(cat), llvm::cl::value_desc("option")} {
+    llvm::errs() << "\n\nConstructed\n\n";
```
Comment (youngar) [inline include/circt/Support/LoweringOptionsParser.h:54]: I think this is debug code

File: lib/Conversion/ExportVerilog/ApplyLoweringOptions.cpp [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+  void setOptions(llvm::StringRef opts) { options = opts.str(); }
+};
+} // namespace
+
+std::unique_ptr<mlir::Pass>
+circt::createTestApplyLoweringOptionPass(llvm::StringRef options) {
+  auto pass = std::make_unique<TestApplyLoweringOptionPass>();
+  pass->setOptions(options);
+  return pass;
+}
```
Comment (youngar) [inline lib/Conversion/ExportVerilog/ApplyLoweringOptions.cpp:47]: I think this is unused?

**Review by youngar — COMMENTED** [review]
Couple more nits

File: lib/Conversion/ExportVerilog/PrepareForEmission.cpp (hunk 2 above)
Comment (darthscsi) [reply lib/Conversion/ExportVerilog/PrepareForEmission.cpp]: yes, this makes a separate pass for testing which can be run from circt-opt to set options on a module.  firtool was doing it before, but it was also being reapplied by exportverilog.

**Pull request merged**